RFCs/0000-pdk-template-v2.md

```diff
@@ -0,0 +1,296 @@
+| tags | An array of string tags for this template. Always includes the directory name of the module. Default is an empty array |
+
+### `<name>/template_schema.json`
+
+### Example templates
+
+#### A template for a default module file
+
+The `.pdkignore` file will be used in all PDK compatible modules therefore it should always be applied.
+
+File layout
+
+``` text
+pdkignore/
+  +- template.yml
+  +- template_schema.json
+  +- files/
+       +- .pdkignore.erb
+```
+
+
+`template.yml`
+
+``` yaml
+---
+name: "PDK Ignore"
+description: >-
+  The PDK Ignore template manages the .pdkignore file for Puppet Modules
+type: module
+always_apply: true
+tags: ignore default
+default_settings:
+  required:
+    - '/appveyor.yml'
+    - '/.fixtures.yml'
+    - '/Gemfile'
+    - '/.gitattributes'
+    - '/.gitignore'
+    - '/.gitlab-ci.yml'
+    - '/.pdkignore'
+    - '/Rakefile'
+    - '/rakelib/'
+    - '/.rspec'
+    - '/.rubocop.yml'
+    - '/.travis.yml'
+    - '/.yardopts'
+    - '/spec/'
+    - '/.vscode/'
+    # ... many other paths
+  paths: ~
+```
+
+`template_schema.json`
+
+``` json
+{
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "$id": "http://puppet.com/schema/does_not_exist.json",
+  "type": "object",
+  "title": "The PDK Ignore Template Schema",
+  "properties": {
+    "paths": {
+      "$id": "#/properties/paths",
+      "title": "Paths property",
+      "description": "Array of file paths to also ignore by the PDK, for example, when building a module",
+      "type": "array",
+      "items": {
+        "type": "string"
+      },
+    },
+  },
+  "definitions": {}
+}
+```
+
+**Note** that the `required` setting is not documented, or has a schema, as it's internal to the template.
+
+`.pdkignore.erb`
+
+Same as [V1 template](https://github.com/puppetlabs/pdk-templates/blob/90e5efab878facd18f838755d9e95b226b1e0776/moduleroot/.pdkignore.erb).
+
+#### An opt-in template for CI configuration
+
+> TODO
+
+#### A template only for new object
+
+
+### Changes to `.sync.yml`
+
+Each module may contain a `.sync.yml` file which contains the module specific settings when rendering the templates. For the new templates, the .sync.yml will have additional settings, which will also be backwards compatible with the V1 templating engine.
+
+* Root `pdk_template` element
+
+  The new `pdk_template` element will hold all of the data for templating engine. Due to being backwards compatible the verbose name of `pdk_template` is used instead of the generic name `template`, so that it would probably not conflict with any known files.  For example, Puppet modules have a `templates` directory.
+
+  * `version` element (Integer)
+
+  Which version of the templating engine this module uses. If this is missing (or the root `pdk_template` element is missing) it is assumed this is Version 1 template. If the template specifies a version that the templating engine does not support it is expected the engine will raise a terminating error.
```

Comment:
I'd be explicit here and say "no `pdk_template` element" == `version: 1`; having a `pdk_template` element means it _has_ to be version 2.

Reply:
Yeah sure.  Will ammend.

Reply:
> having a pdk_template element means it has to be version 2

Could that be tweaked to say that it means it "has to be at least version 2"? Shouldn't we leave ourselves room to revise again down the road if needed?